core/src/main/java/com/linecorp/armeria/server/DefaultServiceConfigSetters.java

```diff
@@ -91,39 +94,60 @@ public ServiceConfigSetters accessLogWriter(AccessLogWriter accessLogWriter, boo
 
     @Override
     public ServiceConfigSetters decorator(Function<? super HttpService, ? extends HttpService> decorator) {
+        return decorator(decorator, DecoratorAndOrder.DEFAULT_ORDER);
+    }
+
+    @Override
+    public ServiceConfigSetters decorator(Function<? super HttpService, ? extends HttpService> decorator,
+                                          int order) {
         requireNonNull(decorator, "decorator");
-        if (this.decorator != null) {
-            this.decorator = this.decorator.andThen(decorator);
-        } else {
-            this.decorator = decorator;
-        }
+        decoratorAndOrders.add(new DecoratorAndOrder<>(decorator, order));
         return this;
     }
 
     Function<? super HttpService, ? extends HttpService> decorator() {
-        if (decorator == null) {
+        final Optional<? extends Function<? super HttpService, ? extends HttpService>> decorator =
+                decoratorAndOrders.stream()
+                                  .sorted((o1, o2) -> Integer.compare(o2.order(), o1.order()))
```

Comment:
How about making `DecoratorAndOrder` implement `Comparable` and just call 
`decoratorAndOrders.stream().sorted()...`?

Reply:
Good idea! 😊